src/Columns/ColumnDecimal.cpp

```diff
@@ -271,7 +271,7 @@ void ColumnDecimal<T>::updatePermutation(IColumn::PermutationSortDirection direc
                 ++index;
             }
 
-            RadixSort<RadixSortTraits<NativeT>>::executeLSD(pairs.data(), size, reverse, res.data());
+            RadixSort<RadixSortTraits<NativeT>>::executeLSD(pairs.data(), size, reverse);
```

Comment:
This fix does work, because now we just sort pairs and does not update result permutation.
This should be proper fix, instead of updating `result` data from beginning we update `begin` iterator:
```
RadixSort<RadixSortTraits<NativeT>>::executeLSD(pairs.data(), size, reverse, begin);
```
I also added test that verify that current fix does not work, and new fix works:
```
SELECT
    (number < 300 ? 0 : 1) AS key,
    (number > 300 ? (number % 2 == 0 ? toDecimal64(number, 8) : toDecimal64(number / 2, 8)) : toDecimal64(0, 8)) AS value
FROM numbers(600)
ORDER BY key, value;
```
You can use this test to verify fix, please consider to add this to functional tests.

Reply:
Indeed. The current fix is the same as NOT SORTING at all. I've also updated the test case that will verify both cases.